Provide a way to specify a minimum log severity level through the CLI.

jonathanknowles opened the pull request:
# Issue Number

#350 

# Overview

This PR provides a way to specify a minimum severity level for log entries through the CLI.

We use the standard taxonomy of [severity levels](https://github.com/input-output-hk/iohk-monitoring-framework/blob/b0ea8317ba5a887d46969e9b3040862a10e6efb3/iohk-monitoring/src/Cardano/BM/Data/Severity.lhs#L33-L44) as provided by [`iohk-monitoring-framework`](https://github.com/input-output-hk/iohk-monitoring-framework):

| **Severity Level** | **Meaning** |
| -- | -- |
| Debug | detailed information about values and decision flow |
| Info      | general information of events; progressing properly |
| Notice    | needs attention; something not progressing properly |
| Warning   | may continue into an error condition if continued |
| Error     | unexpected set of event or condition occurred |
| Critical  | error condition causing degrade of operation |
| Alert     | a subsystem is no longer operating correctly, likely requires manual intervention |
| Emergency | at this point, the system can never progress without additional intervention |

I have:

- [x] Added a CLI option `--min-log-severity`
- [x] Added a default log severity level of `debug` (which will include everything)
- [x] Used the specified log severity level to configure the `Tracer` object at startup.
- [x] Provided a round-trip test to ensure that severity levels can be converted to and from `Text`. 
- [x] Tested (informally) that log entries are indeed filtered out if they are below the specified minimum severity level, but included if they are greater than or equal in severity.
- [x] Enumerate the set of expected values in the error message on failure to decode a Severity level from `Text`.

## Further work

It would be desirable to have an automated test to ensure that we filter out all log entries that are less severe than the specified minimum severity level (and that we retain those entries greater or equal in severity to the specified level).

jonathanknowles committed:
Provide a way to specify a minimum log severity level through the CLI.

We use the standard taxonomy of severity levels as provided by `iohk-monitoring-framework`

KtorZ commented on exe/wallet/Main.hs:
```diff
@@ -182,9 +189,10 @@ main = do
     c <- defaultConfigStdout
+    setMinSeverity c minSeverity
     tr <- setupTrace (Right c) "simple"
     appendName "text" tr
```
Not related to this PR but we should review the `"simple"` and `"text"` too here, since logged currently gets displayed as: `[iohk.simple.text] <log message>` ... Not sure we need to append anything, and `simple` could just be `cardano-wallet` I believe :smile:

jonathanknowles replied:
> I wonder if and when we could have a configuration file at some point. I admit I don't like the idea too much and would prefer having everything in CLI and/or in the ENV.

The number of options is certainly becoming longer, and it certainly makes the help text unwieldy to look at. Right now, we display **all options** for **all commands** in the help text.

Perhaps we could alleviate this by using a different CLI parsing framework that has a nicer way to display the various options?

With a different framework, we could display a general overview when the user types `--help`, and then give greater levels of detail when the user asks for help in a specific area (for example `cardano-wallet mnemonic --help`) or for a specific command, like `mnemonic generate --help`.

KtorZ replied:
Yes! And that is already planned :) Most likely coming for our next sprint.

jonathanknowles replied:
> Not related to this PR but we should review the `"simple"` and `"text"` too here, since logged currently gets displayed as: `[iohk.simple.text] <log message>` ... Not sure we need to append anything, and `simple` could just be `cardano-wallet` I believe

Great point! I've added a task to do this.

KtorZ commented on lib/cli/src/Cardano/CLI.hs:
```diff
@@ -106,10 +109,39 @@ instance FromText (Port tag) where
+        "warning" ->
+            mk Warning
+        _ -> Left $ TextDecodingError $ show (T.unpack t)
+            <> " is not a recognized log severity level."
```
What about mentioning the expected values in the error message. It makes much more user-friendly errors. "log severity level is neither alert, criticial, debug .... nor warning (maybe ordering them by severity!)

jonathanknowles replied:
That's a sensible suggestion! I've added a task to this PR.

jonathanknowles replied:
@KtorZ I've added a patch to enumerate the complete list of severity levels within the error message. They're listed in ascending order of severity.

KtorZ reviewed (approved):
I wonder if and when we could have a configuration file at some point. I admit I don't like the idea too much and would prefer having everything in CLI and/or in the ENV.

jonathanknowles committed:
Remove extraneous vertical space.

https://github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards#use-only-a-single-blank-line-between-top-level-definitions

jonathanknowles committed:
Enumerate all `Severity` levels within `TextDecodingError`.

On failure to decode a `Severity` level from `Text`, enumerate all valid
`Severity` levels in the error message.

rvl reviewed (approved):
It looks like it will work.

jonathanknowles committed:
Make it clear that log severity levels are listed in ascending order.
